pyfr/integrators/dual/pseudo/multip.py

```diff
@@ -0,0 +1,248 @@
+        if l1 != self._order:
+            self.integrator._add(1, rtemp, -1, self._mg_regidx[0])
+
+        # Activate l2 system and get l2 regidx
+        self.level = l2
+        mg0, mg1 = self._mg_regidx
+
+        # Restrict Q
+        l1sys.eles_scal_upts_inb.active = l1idxcurr
+        l2sys.eles_scal_upts_inb.active = l2idxcurr
+        self.integrator._queue % self.mgproject(l1, l2)()
+
+        # Restrict d and store to mg1
+        l1sys.eles_scal_upts_inb.active = rtemp
+        l2sys.eles_scal_upts_inb.active = mg1
+        self.integrator._queue % self.mgproject(l1, l2)()
+
+        # mg0 = R = -∇·f - dQ/dt
+        self.integrator.system.rhs(self.tcurr, l2idxcurr, self._mg_regidx[0])
+
+        # Compute the target residual r
+        # mg0 = r = R + d
+        self.integrator._add(1, self._mg_regidx[0], -1, self._mg_regidx[1])
+
+        # Need to store the non-smoothed solution Q^ns for the correction
+        # mg1 = Q^ns
+        self.integrator._add(0, mg1, 1, l2idxcurr)
+
+        # Restrict the dt source terms
+        for i in range(len(self.integrator._source_regidx)):
+            l1sys.eles_scal_upts_inb.active = self.integrators[l1]._source_regidx[i]
+            l2sys.eles_scal_upts_inb.active = self.integrators[l2]._source_regidx[i]
+            self.integrator._queue % self.mgproject(l1, l2)()
+
+    def prolongate(self, l1, l2):
+        l1idxcurr  = self.integrators[l1]._idxcurr
+        l2idxcurr = self.integrators[l2]._idxcurr
+
+        l1sys, l2sys = self.integrators[l1].system, self.integrators[l2].system
+
+        # Prevsoln is used as temporal storage at l2
+        rtemp = 0 if l2idxcurr == 1 else 1
+
+        # Correction with respect to the non-smoothed value from down-cycle
+        # mg1 = Delta = Q^s - Q^ns
+        self.integrator._add(-1, self._mg_regidx[1], 1, l1idxcurr)
+
+        # Prolongate the correction and store to rtemp
+        l1sys.eles_scal_upts_inb.active = self._mg_regidx[1]
+        l2sys.eles_scal_upts_inb.active = rtemp
+        self.integrator._queue % self.mgproject(l1, l2)()
+
+        # Add the correction to the end quantity at l2
+        # Q^m+1  = Q^s + Delta
+        self.level = l2
+        self.integrator._add(1, l2idxcurr, 1, rtemp)
+
+    @property
+    def _mg_regidx(self):
+        if self.level == self._order:
+            raise AttributeError('_mg_regidx not defined when'
+                                 ' self.level == self._order')
+
+        return self.integrator._regidx[-2:]
+
+    def pseudo_advance(self, tcurr, tout, dt):
+        # Multigrid levels and step counts
+        cycle, csteps = self.cycle, self.csteps
+
+        self.tcurr = tcurr
+
+        for i in range(self._maxniters):
+            for l, m, n in it.zip_longest(cycle, cycle[1:], csteps):
+                self.level = l
+                # Set the number of smoothing steps at each level
+                self.integrator._maxniters = self.integrator._minniters = n
```

Comment:
These should probably be non ``_`` if someone else is setting them outside the class.